src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs

```diff
@@ -441,6 +443,212 @@ public List<PackageManagerSearchElement> ListAll()
             return CachedPackageList;
         }
 
+        internal void InitiatePackageDownloadAndInstall(IPackageInfo packageInfo, string downloadPath = null)
+        {
+            var header = Model.GetPackageHeader(packageInfo);
+            if (header == null)
+            {
+                // TODO: move message to resources
+                var result = MessageBox.Show(string.Format("{0} could not be downloaded.", packageInfo.Name),
+                "Download Error", MessageBoxButton.OKCancel, MessageBoxImage.Question);
+                return;
+            }
+
+            var version = Model.GetGregPackageVersion(header, packageInfo.Version);
+            if (version == null)
+            {
+                // TODO: move message to resources
+                var result = MessageBox.Show(string.Format("Version {0} of {1} could not be found.", packageInfo.Version.ToString(), packageInfo.Name),
+                "Download Error", MessageBoxButton.OKCancel, MessageBoxImage.Question);
+                return;
+            }
+
+            ExecutePackage(packageInfo.Name, version, downloadPath);
+        }
+
+        private string JoinPackageNames(IEnumerable<Package> pkgs)
+        {
+            return String.Join(", ", pkgs.Select(x => x.Name + " " + x.VersionName));
+        }
+
+        internal void ExecutePackage(string name, PackageVersion version, string downloadPath)
```

Comment:
what does this name mean?

Reply:
@scottmitchell You probably mean `DownloadPackage` here?

Reply:
Most of this logic was previously in a method called `PackageOnExecuted` on `PackageManagerSearch` before I moved it to the `PackageManagerClient` in order for it to be accessed from other extensions. The name was a vestige of that. I've changed it to `ExecutePackageDownload`.